Add "Optimizing Build Performance" section to the Cargo book (#15924)
rust-lang/cargo: src/doc/src/guide/build-performance.md

Reviewer: I was considering bringing up dividing this into two sections
- Root causing build times (`--timings`, compiler passes, macro lines, `cargo llvm-lines`)
- Optimizing builds

I was just split on that vs framing those in terms of the current organization.  For example, we could have a section on reducing the impact of generics and talk through using `cargo llvm-lines` to find large, repeated generic functions and discuss how to improve that.

# Optimizing Build Performance

Compilation of Rust crates can sometimes be rather slow, due to various reasons (such as the used compilation model and the design of the Rust language and its compiler). There are various approaches that can be used to optimize build performance, which mostly fall under two categories:

- Modify compiler or Cargo flags
- Modify the source code of your crate(s)

This guide focuses on the first approach.

> Below, you can find several methods that can be used to optimize build performance. It is important to nore that their effect varies a lot based on the compiled crate, and in some cases they can actually make compilation slower. You should always measure build performance on your crate(s) to determine if a given method described here is effective for your crate.

Reviewer 2: ```suggestion
Below, you can find several methods that can be used to optimize build performance. It is important to note that their effect varies a lot based on the compiled crate, and in some cases they can actually make compilation slower. You should always measure build performance on your crate(s) to determine if a given method described here is effective for your crate.
```

Note that some of these approaches currently require using the nightly toolchain.

> ## Reduce amount of generated debug information

Reviewer: For a section like this, we've been talking about making changes.  Unsure if there is an issue (and I'm generally cautious about linking to issues in documentation) but I feel like doing so could help gather more insight (e.g. the iteration vs debugging trade off) and give people a sense of "things are moving in a positive direction"

> ## Reduce amount of generated debug information

Reviewer 3: > I wrote up an approach that applies those ideas to this section at Kobzol#1.

I like the writing style listing tradeoff.

And agree it is overlapped with other reference-like doc.

[profile.dev]
debug = false # or "line-tables-only"
```

If you want to keep debuginfo for your crate only, but you do not need it for your dependencies, you can set `debug = false` as the default value for a given profile, and then enable debuginfo only for your crate:

```toml
[profile.dev]
debug = false

[profile.dev.package]
<your-crate-name>.debug = true
```

## Use an alternative codegen backend

> **This requires nightly/unstable features**

The component of the Rust compiler that generates executable code is called a "codegen backend". The default backend is LLVM, which produces very optimized code, at the cost of relatively slow compilation time. You can try to use a different codegen backend in order to speed up the compilation of your crate.

You can use the [Cranelift](https://github.com/rust-lang/rustc_codegen_cranelift) backend, which is designed for fast(er) compilation time. You can install this backend using rustup:

```console
$ rustup component add rustc-codegen-cranelift-preview --toolchain nightly
```

and then enable it for a given Cargo profile using the `codegen-backend` option in `Cargo.toml`:
```toml
[profile.dev]
codegen-backend = "cranelift"
```

> Since this is currently an unstable option, you will also need to either pass `-Z codegen-backend` to Cargo, or enable this unstable option in the `.cargo/config.toml` file. You can find more information about the unstable `codegen-backend` profile option [here](../reference/unstable.md#codegen-backend).

Reviewer 3: Instead of here, should we do this?

```suggestion
Since this is currently an unstable option, you will also need to either pass `-Z codegen-backend` to Cargo, or enable this unstable option in the `.cargo/config.toml` file. You can find more information in the unstable [`codegen-backend` profile option](../reference/unstable.md#codegen-backend) section.
```

# Optimizing Build Performance

> Cargo uses configuration defaults that try to balance various aspects, including debuggability, runtime performance, build performance, binary size and others. Because of that, build performance is sometimes traded off for other benefits which may not be as important for your circumstances. This guide will step you through changes you can make to improve build performance.

Author: I see. My original thought was that this page should mostly talk about Cargo options, but if you want to expand that to go "all the way", and also give hints about splitting crates, reducing dependency counts, reducing usage of generics/macros, then it of course makes sense.

In the compiler performance survey, we got quite a lot of open responses that said that people don't want to go change their code to optimize build performance at all, which is understandable (although not very realistic with the current state of things).

> Cargo uses configuration defaults that try to balance various aspects, including debuggability, runtime performance, build performance, binary size and others. Because of that, build performance is sometimes traded off for other benefits which may not be as important for your circumstances. This guide will step you through changes you can make to improve build performance.

Reviewer: I figure it would be good to have things in one place vs having cargo optimizations here and code changes elsewhere.  We similarly have a SemVer compatibility page that covers Cargo and Rust.

> In the compiler performance survey, we got quite a lot of open responses that said that people don't want to go change their code to optimize build performance at all, which is understandable (although not very realistic with the current state of things).

They can always skip those sections :)

The component of the Rust compiler that generates executable code is called a "codegen backend". The default backend is LLVM, which produces very optimized code, at the cost of relatively slow compilation time. You can try to use a different codegen backend in order to speed up the compilation of your crate.

You can use the [Cranelift](https://github.com/rust-lang/rustc_codegen_cranelift) backend, which is designed for fast(er) compilation time. You can install this backend using rustup:

```console
$ rustup component add rustc-codegen-cranelift-preview --toolchain nightly
```

and then enable it for a given Cargo profile using the `codegen-backend` option in `Cargo.toml`:
```toml
[profile.dev]
codegen-backend = "cranelift"
```

Since this is currently an unstable option, you will also need to either pass `-Z codegen-backend` to Cargo, or enable this unstable option in the `.cargo/config.toml` file. You can find more information about the unstable `codegen-backend` profile option [here](../reference/unstable.md#codegen-backend).

Note that the Cranelift backend might not support all features used by your crate. It is also available only for a limited set of targets.

> ## Where to apply configuration changes

Reviewer: This feels weird to be at the same header level as the rest.  Unsure what we typically do

# Optimizing Build Performance

> This guide will step you through Cargo configuration options and source code organization patterns that can help improve build performance, by prioritizing it over other aspects which may not be as important for your circumstances.

Reviewer: ```suggestion
Cargo configuration options and source code organization patterns can help improve build performance, by prioritizing it over other aspects which may not be as important for your circumstances.
```

### Reduce amount of generated debug information

Recommendation: Add to your `Cargo.toml` or `.cargo/config.toml`:

```toml
[profile.dev]
debug = "line-tables-only"

[profile.dev.package."*"]
debug = false

[profile.debugging]
inherits = "dev"
debug = true
```

This will:
- Change the [`dev` profile](../reference/profiles.md#dev) (default for development commands) to:
  - Limit [debug information](../reference/profiles.md#debug) for workspace members to what is needed for useful panic backtraces
  - Avoid generating any debug information for dependencies
- Provide an opt-in for when debugging via [`--profile debugging`](../reference/profiles.md#custom-profiles)

Trade-offs:
- ✅ Faster build times
- ✅ Faster link times
- ✅ Smaller disk usage of the `target` directory
- ❌ Requires a full rebuild to have a high-quality debugger experience

### Use an alternative codegen backend

> > **This requires nightly/unstable features**

Author: It is, but I thought that the nightly/unstable aspect of it is the most important thing to consider, that is also why I have been accenting it so much in the guide. IMO nightly will be a show-stopper for many people and listing it just as another trade-off feels wrong.

Recommendation:

- Install the Cranelift codegen backend rustup component
    ```console
    $ rustup component add rustc-codegen-cranelift-preview --toolchain nightly
    ```
- Add to your `Cargo.toml` or `.cargo/config.toml`:
    ```toml
    [profile.dev]
    codegen-backend = "cranelift"
    ```
- Run Cargo with `-Z codegen-backend` or enable the [`codegen-backend`](../reference/unstable.md#codegen-backend) feature in `.cargo/config.toml`.
  - This is required because this is currently an unstable feature.

This will change the [`dev` profile](../reference/profiles.md#dev) to use the [Cranelift codegen backend](https://github.com/rust-lang/rustc_codegen_cranelift) for generating machine code, instead of the default LLVM backend. The Cranelift backend should generate code faster than LLVM, which should result in improved build performance.

Trade-offs:
- ✅ Faster build times

> - ❌ Worse runtime performance of the generated code

Reviewer: If I'm understanding correctly, being more specific, this is
- This always speeds up builds of proc macros and build scripts
- For everything else, this only helps with `cargo build` and not `cargo check`
- `cargo build` will mostly be done through `cargo test` and this could make `cargo test` times worse than the savings in `cargo build`

Any way we could provide better nuance here?  Saying "faster build times" may sound more broad than it is.  And saying worse runtime performance, people may not pick up on the implications.

- ❌ Requires using nightly Rust and an unstable feature
- ❌ Only available for [certain targets](https://github.com/rust-lang/rustc_codegen_cranelift?tab=readme-ov-file#platform-support)
- ❌ Might not support all Rust features (e.g. unwinding)

## Source code organization

> TODO

Reviewer: Even if we leave merge this after beta branch, I'd rather not be leaving this TODO section in
